cmd/bench/cmd/node.go

```diff
@@ -0,0 +1,182 @@
+// Copyright Contributors to the Mir project
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package cmd
+
+import (
+	"context"
+	"encoding/csv"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"github.com/filecoin-project/mir"
+	mirCrypto "github.com/filecoin-project/mir/pkg/crypto"
+	"github.com/filecoin-project/mir/pkg/iss"
+	"github.com/filecoin-project/mir/pkg/logging"
+	"github.com/filecoin-project/mir/pkg/modules"
+	"github.com/filecoin-project/mir/pkg/net"
+	"github.com/filecoin-project/mir/pkg/net/grpc"
+	"github.com/filecoin-project/mir/pkg/net/libp2p"
+	"github.com/filecoin-project/mir/pkg/requestreceiver"
+	t "github.com/filecoin-project/mir/pkg/types"
+	grpctools "github.com/filecoin-project/mir/pkg/util/grpc"
+	libp2ptools "github.com/filecoin-project/mir/pkg/util/libp2p"
+)
+
+const (
+	NodeBasePort        = 10000
+	ReqReceiverBasePort = 20000
+)
+
+var (
+	transportType string
+	statFileName  string
+	statPeriod    time.Duration
+
+	nodeCmd = &cobra.Command{
+		Use:   "node",
+		Short: "Start a Mir node",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runNode()
+		},
+	}
+)
+
+func init() {
+	rootCmd.AddCommand(nodeCmd)
+	nodeCmd.Flags().StringVarP(&transportType, "net", "t", "libp2p", "network transport (libp2p|grpc)")
+	nodeCmd.Flags().StringVarP(&statFileName, "statFile", "o", "", "output file for statistics")
+	nodeCmd.Flags().DurationVar(&statPeriod, "statPeriod", time.Second, "statistic record period")
+}
+
+func runNode() error {
+	var logger logging.Logger
+	if verbose {
+		logger = logging.ConsoleDebugLogger
+	} else {
+		logger = logging.ConsoleWarnLogger
+	}
+
+	ownID, err := strconv.Atoi(id)
+	if err != nil {
+		return fmt.Errorf("unable to convert node ID: %w", err)
+	} else if ownID < 0 || ownID >= nrNodes {
+		return fmt.Errorf("ID must be in [0, %d]", nrNodes-1)
+	}
+
+	nodeIDs := make([]t.NodeID, nrNodes)
+	for i := range nodeIDs {
+		nodeIDs[i] = t.NewNodeIDFromInt(i)
+	}
+
+	reqReceiverAddrs := make(map[t.NodeID]string)
+	for i := range nodeIDs {
+		reqReceiverAddrs[t.NewNodeIDFromInt(i)] = fmt.Sprintf("127.0.0.1:%d", ReqReceiverBasePort+i)
+	}
+
+	var transport net.Transport
+	nodeAddrs := make(map[t.NodeID]t.NodeAddress)
+	switch strings.ToLower(transportType) {
+	case "grpc":
```

Comment:
Depending on what we want to measure, local grpc/libp2p transport may not make much sense (there will be no network delays, only marshaling/unmarshaling delays).

Reply:
The tool would be enhanced with the ability to run on different machines.

Reply:
It seems like now we can do that with the membership configuration file.

Reply:
Btw example membership config files are already in the chat demo. I guess I should have created an example directly within the bench folder...